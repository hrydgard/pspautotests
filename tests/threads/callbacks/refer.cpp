#include "shared.h"

int cbFunc(int arg1, int arg2, void *arg) {
	return 0;
}

void testRefer(const char *title, SceUID cb, SceKernelCallbackInfo *info) {
	int result = sceKernelReferCallbackStatus(cb, info);
	if (result == 0 && info != NULL) {
		checkpoint(NULL);
		schedf("%s: ", title);
		schedfCallback(*info);
	} else {
		checkpoint("%s: Failed (%08x)", title, result);
	}
}

extern "C" int main(int argc, char *argv[]) {
	SceUID cb = sceKernelCreateCallback("refer", &cbFunc, (void *)0xDEADBEEF);
	SceKernelCallbackInfo info;
	info.size = sizeof(info);

	// Crashes.
	//testRefer("NULL info", cb, NULL);
	testRefer("Normal", cb, &info);

	checkpointNext("Sizes:");
	int sizes[] = {0, 1, 2, 4, 8, 44, 48, 52, 56, 60, 64, 68, 72, 128, 256, 1024, 0xFF00, 0xFF0000, 0x80000001, 0xCCCCCCCC, 0xFFFFFFFF};
	int i, result;
	for (i = 0; i < sizeof(sizes) / sizeof(sizes[0]); ++i) {
		info.size = sizes[i];
		result = sceKernelReferCallbackStatus(cb, &info);
		checkpoint("  %08x => %08x (result=%08x)", sizes[i], info.size, result);
	}

	sceKernelDeleteCallback(cb);

	info.size = sizeof(info);
	checkpointNext("Objects:");
	testRefer("NULL", 0, &info);
	testRefer("Invalid", 0xDEADBEEF, &info);
	testRefer("Deleted", cb, &info);

	return 0;
}