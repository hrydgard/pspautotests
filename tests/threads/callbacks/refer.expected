[x] Normal: Callback: OK (size=56,name=refer,thread=1,callback=1,common=deadbeef,notifyCount=00000000,notifyArg=0)

[x] Sizes:
[x]   00000000 => 00000000 (result=00000000)
[x]   00000001 => 00000038 (result=00000000)
[x]   00000002 => 00000038 (result=00000000)
[x]   00000004 => 00000038 (result=00000000)
[x]   00000008 => 00000038 (result=00000000)
[x]   0000002c => 00000038 (result=00000000)
[x]   00000030 => 00000038 (result=00000000)
[x]   00000034 => 00000038 (result=00000000)
[x]   00000038 => 00000038 (result=00000000)
[x]   0000003c => 00000038 (result=00000000)
[x]   00000040 => 00000038 (result=00000000)
[x]   00000044 => 00000038 (result=00000000)
[x]   00000048 => 00000038 (result=00000000)
[x]   00000080 => 00000038 (result=00000000)
[x]   00000100 => 00000038 (result=00000000)
[x]   00000400 => 00000038 (result=00000000)
[x]   0000ff00 => 00000038 (result=00000000)
[x]   00ff0000 => 00000038 (result=00000000)
[x]   80000001 => 00000038 (result=00000000)
[x]   cccccccc => 00000038 (result=00000000)
[x]   ffffffff => 00000038 (result=00000000)

[x] Objects:
[x] NULL: Failed (800201a1)
[x] Invalid: Failed (800201a1)
[x] Deleted: Failed (800201a1)
